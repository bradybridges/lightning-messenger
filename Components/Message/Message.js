import React, { Component } from 'react';
import { Text, View, StyleSheet, Dimensions } from 'react-native';

export default class Message extends Component {
  formatTimestamp = (timestamp) => {
    const date = new Date(timestamp.seconds * 1000);
    let hours = Number(date.getHours());
    const label = (hours >= 12) ? 'PM': 'AM';
    let minutes = Number(date.getMinutes());
    if(hours > 12) {
      hours = hours - 12;
    }
    if(minutes < 10) {
      minutes = `0${minutes}`;
    }
    return `${hours}:${minutes} ${label}`;
  }
  render() {
    const { content, timestamp } = this.props;
    const time = this.formatTimestamp(timestamp);
    return (
      <View style={styles.container}>
        <Text style={styles.content}>{content}</Text>
        <Text style={styles.timestamp}>{time}</Text>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    alignItems: 'center',
    backgroundColor: 'red',
    borderRadius: 100,
    marginVertical: 30,
  },
  content: {
    padding: 10,
    fontSize: 18,
  },
  timestamp: {
    alignSelf: 'flex-end',
    marginRight: 20,
  }
});

