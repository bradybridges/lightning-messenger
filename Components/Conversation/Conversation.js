import React, { Component } from 'react';
import { 
  Text,
  View,
  StyleSheet,
  Dimensions,
  ScrollView
} from 'react-native';
import Message from '../Message/Message';
import ComposeMessageForm from '../ComposeMessageForm/ComposeMessageForm';
import BackButton from '../BackButton/BackButton';
import * as Constants from '../../Constants/Constants';
import { getStatusBarHeight } from 'react-native-status-bar-height';

export default class Conversation extends Component {
  renderMessages = () => {
    const { messages, deleteMessage } = this.props;
    return messages.map((message, i) => {
      return (
        <Message 
          key={i} 
          content={message.contents} 
          timestamp={message.timestamp}
          isSender={message.sender === true ? true: false}
          deleteMessage={this.handleDeleteMessage}
        />
      );  
    });
  }

  handleDeleteMessage = async (content, isSender) => {
    const { deleteMessage, from } = this.props;
    await deleteMessage(from, content, isSender);
  }

  render() {
    const { from, messages, updateConversation, closeSelectedConversation } = this.props;
    const { email } = this.props.user;
    return (
      <View style={styles.container}>
        <View style={styles.headerContainer}>
          <BackButton close={closeSelectedConversation} />
          <Text style={styles.header}>{from}</Text>
        </View>
        <View style={styles.messagesContainer}>
          <ScrollView contentContainerStyle={{ display: 'flex' }}>
            { this.renderMessages() }
          </ScrollView>
        </View>
        <ComposeMessageForm from={email} to={from} updateConversation={updateConversation}/>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    backgroundColor: Constants.primaryBgColor,
    justifyContent: 'flex-start',
    paddingTop: Dimensions.get('window').height * .1,
  },
  headerContainer: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    height: getStatusBarHeight(true) + 50,
    backgroundColor: Constants.primaryHeaderColor,
    padding: Constants.baseMarginPadding,
    position: 'absolute',
    zIndex: 5,
    width: '100%',
    paddingTop: getStatusBarHeight(true),
  },
  header: {
    color: 'white',
    fontSize: 24,
    width: '90%',
    textAlign: 'center',
    fontFamily: 'exo-regular',
  },
  messagesContainer: {
    width: Dimensions.get('window').width,
    height: Dimensions.get('window').height * .8,
  }
});

