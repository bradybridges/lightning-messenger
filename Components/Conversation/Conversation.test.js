import React from 'react';
import { shallow } from 'enzyme';
import Conversation from './Conversation';
import Message from '../Message/Message';
import { mockMessages } from '../../mockMessages';

describe('Conversation', () => {
  let wrapper;
  const mockUpdateConverstaion = jest.fn();
  const mockUser = { email: 'test@test.com' };
  const from = 'yourfriend@test.com';
  beforeEach(() => {
    wrapper = shallow(
      <Conversation 
        from={from}
        user={mockUser}
        messages={mockMessages}
        updateConversation={mockUpdateConverstaion}
      />
    );
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renderMessages should return an array of Message components', () => {
    const messages = wrapper.instance().renderMessages();
    expect(messages.length).toEqual(3);
  });
});