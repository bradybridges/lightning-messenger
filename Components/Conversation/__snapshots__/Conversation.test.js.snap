// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conversation should match snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000000d6",
      "flex": 1,
      "justifyContent": "flex-start",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#00000069",
        "display": "flex",
        "height": 133.4,
        "justifyContent": "center",
        "padding": 10,
        "position": "absolute",
        "width": "100%",
        "zIndex": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 24,
        }
      }
    >
      yourfriend@test.com
    </Text>
  </View>
  <View
    style={
      Object {
        "height": 1200.6000000000001,
        "width": 750,
      }
    }
  >
    <ScrollViewMock>
      <Message
        content="test msg"
        key="0"
        timestamp={
          Object {
            "seconds": "1581031140",
          }
        }
      />
      <Message
        content="good day"
        key="1"
        timestamp={
          Object {
            "seconds": "1581031189",
          }
        }
      />
      <Message
        content="test response"
        key="2"
        timestamp={
          Object {
            "seconds": "1581031231",
          }
        }
      />
    </ScrollViewMock>
  </View>
  <ComposeMessageForm
    from="test@test.com"
    to="yourfriend@test.com"
    updateConversation={[MockFunction]}
  />
</View>
`;
